File: Entity Framework in Windows forms part 2/[C#]-Entity Framework in Windows forms part 2/C#/Example1_cs/LanguageExtensions/Extensions.cs
using EntityFrameWorkNorthWind_cs;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;

namespace Example1_cs
{
    public static class Extensions
    {
        /// <summary>
        /// Check if BindingSource.Current has data
        /// </summary>
        /// <param name="sender"></param>
        /// <returns></returns>
        public static bool IsValid(this Customer sender)
        {
            return (sender != null);
        }
        /// <summary>
        /// Determines if the current item is not null
        /// </summary>
        /// <param name="sender"></param>
        /// <returns></returns>
        public static bool CurrentRowIsValid(this BindingSource sender)
        {
            return (sender.Current != null);
        }
        /// <summary>
        /// Returns Customer for the current property
        /// </summary>
        /// <param name="sender"></param>
        /// <returns></returns>
        public static Customer Customer(this BindingSource sender)
        {
            return ((Customer)sender.Current);
        }
        /// <summary>
        /// Get names of an Enum
        /// </summary>
        /// <param name="sender">Type of Enum (see example)</param>
        /// <returns>Names of members in Enum</returns>
        public static string[] Names(this Type sender)
        {
            return Enum.GetNames(sender);
        }

        /// <summary>
        /// Converts string value to a member of an Enum
        /// </summary>
        /// <typeparam name="T">Valid Enum structure</typeparam>
        /// <param name="sender">String to convert</param>
        /// <returns>A member of the enum</returns>
        /// <example>
        /// <code source="CodeExamples\EnumExamples.vb" language="vbnet" title="VB.NET Examples"/>
        /// </example>
        public static T ToEnum<T>(this string sender)
        {
            Type senderType = typeof(T);

            if (Enum.IsDefined(typeof(T), sender))
            {
                return (T)Enum.Parse(senderType, sender, true);
            }
            else
            {
                string baseType = senderType.ToString();
                int position = baseType.IndexOf('+');
                string errorMessage = "";

                if (position > -1)
                {
                    string EnumName = baseType.Substring(position + 1);
                    errorMessage = string.Format("'{0}' not a member of '{1}'", sender, EnumName);
                }
                else
                {
                    errorMessage = string.Format("{0} not a member of {1}", sender, senderType);
                }

                throw new Exception(errorMessage);

            }
        }
        /// <summary>
        /// Used to filter data via a generic object using lambda
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="list"></param>
        /// <param name="filterParam"></param>
        /// <returns></returns>
        public static IEnumerable<T> Filter<T>(this IEnumerable<T> list, Func<T, bool> filterParam)
        {
            return list.Where(filterParam);
        }
        /// <summary>
        /// Expand all columns excluding in this case Orders column
        /// </summary>
        /// <param name="sender"></param>
        public static void ExpandColumns(this DataGridView sender)
        {
            foreach (DataGridViewColumn col in sender.Columns)
            {
                // ensure we are not attempting to do this on a Entity
                if (col.ValueType.Name != "ICollection`1")
                {
                    col.AutoSizeMode = DataGridViewAutoSizeColumnMode.AllCells;
                }
            }
        }
        /// <summary>
        /// Provides functionality to get a distinct list of items used in a lambda Distint call.
        /// </summary>
        /// <typeparam name="TSource"></typeparam>
        /// <typeparam name="TKey"></typeparam>
        /// <param name="sender"></param>
        /// <param name="keySelector"></param>
        /// <returns></returns>
        public static IEnumerable<TSource> DistinctBy<TSource, TKey>(this IEnumerable<TSource> sender, Func<TSource, TKey> keySelector)
        {
            HashSet<TKey> knownKeys = new HashSet<TKey>();

            foreach (TSource element in sender)
            {
                if (knownKeys.Add(keySelector(element)))
                {
                    yield return element;
                }
            }
        }
    }
}

File: Entity Framework 6 for VB.NET Windows forms solutions/[C#][VB]-Entity Framework 6 for VB.NET Windows forms solutions/C# and VB/BackendOperations/Extensions/Extensions.vb
Public Module Extensions
    ''' <summary>
    ''' Maps CustomerDTO to Customer
    ''' </summary>
    ''' <param name="CustomerDto"></param>
    ''' <returns></returns>
    ''' <remarks>
    ''' For complex mapping, our team uses AutoMapper which
    ''' is a complete overkill for here.
    ''' </remarks>
    <Runtime.CompilerServices.Extension()>
    Public Function ToCustomer(ByVal CustomerDto As CustomerDTO) As Customer
        Dim Customer As New Customer With
                {
                    .id = CustomerDto.id,
                    .FirstName = CustomerDto.FirstName,
                    .LastName = CustomerDto.LastName,
                    .Address = CustomerDto.Address,
                    .City = CustomerDto.City,
                    .State = CustomerDto.State,
                    .ZipCode = CustomerDto.ZipCode
                }

        Return Customer

    End Function
    <Runtime.CompilerServices.Extension()>
    Public Function ToCustomerDTO(ByVal Customer As Customer) As CustomerDTO
        Dim CustomerDto As New CustomerDTO With
            {
                .id = Customer.id,
                .FirstName = Customer.FirstName,
                .LastName = Customer.LastName,
                .Address = Customer.Address,
                .City = Customer.City,
                .State = Customer.State,
                .ZipCode = Customer.ZipCode
            }

        Return CustomerDto

    End Function

    ''' <summary>
    ''' Used to filter data via a generic object using lambda
    ''' </summary>
    ''' <typeparam name="T"></typeparam>
    ''' <param name="list"></param>
    ''' <param name="filterParam"></param>
    ''' <returns></returns>
    <Runtime.CompilerServices.Extension>
    Public Function Filter(Of T)(ByVal list As IEnumerable(Of T), ByVal filterParam As Func(Of T, Boolean)) As IEnumerable(Of T)
        Return list.Where(filterParam)
    End Function
    ''' <summary>
    ''' Converts string value to a member of an Enum
    ''' </summary>
    ''' <typeparam name="T">Valid Enum structure</typeparam>
    ''' <param name="sender">String to convert</param>
    ''' <returns>A member of the enum</returns>
    ''' <example>
    ''' <code source="CodeExamples\EnumExamples.vb" language="vbnet" title="VB.NET Examples"/>
    ''' </example>
    <System.Runtime.CompilerServices.Extension>
    Public Function ToEnum(Of T)(ByVal sender As String) As T
        Dim senderType As Type = GetType(T)

        If System.Enum.IsDefined(GetType(T), sender) Then
            Return CType(System.Enum.Parse(senderType, sender, True), T)
        Else
            Dim baseType As String = senderType.ToString()
            Dim position As Integer = baseType.IndexOf("+"c)
            Dim errorMessage As String = ""

            If position > -1 Then
                Dim EnumName As String = baseType.Substring(position + 1)
                errorMessage = String.Format("'{0}' not a member of '{1}'", sender, EnumName)
            Else
                errorMessage = String.Format("{0} not a member of {1}", sender, senderType)
            End If

            Throw New Exception(errorMessage)

        End If
    End Function

End Module
